Add moderation-event GET endpoint

VadimKovalenkoSNF opened the pull request:
Fix [OSDEV-1346](https://opensupplyhub.atlassian.net/browse/OSDEV-1346)

Add GET `api/v1/moderation-events`

roman-stolar commented:
> ## [![Quality Gate Failed](https://camo.githubusercontent.com/ce402bcc4cf818eda660128f0b27a891eb8c9f4e0d783303ea667cb0d4b07f95/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=opensupplyhub_open-supply-hub&pullRequest=391) **Quality Gate failed**
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [7 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=opensupplyhub_open-supply-hub&pullRequest=391&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=opensupplyhub_open-supply-hub&pullRequest=391)

@VadimKovalenkoSNF If you disagree with those security fails from SonarQube you can resolved it redirecting to SonarQube service.

VadimKovalenkoSNF commented:
TODO: redo **Match Status** to **PENDING**,  **APPROVED**, **REJECTED**.

VadimKovalenkoSNF commented on src/django/api/migrations/0159_alter_status_of_moderation_event_table.py:
```diff
@@ -0,0 +1,19 @@
+        ('api', '0158_create_moderation_event_table'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='moderationevent',
+            name='status',
+            field=models.CharField(choices=[('PENDING', 'Pending'), ('APPROVED', 'Approved'), ('REJECTED', 'Rejected')], default='PENDING', help_text='Moderation status of the production location.', max_length=8),
+        ),
```
We don't have related data in DB yet.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/moderation_events_query_builder.py:
```diff
@@ -0,0 +1,103 @@
+
+        if field == V1_PARAMETERS_LIST.OS_ID:
+            self._build_os_id(values)
+
```
It is called from a parent class interface.

VadimKovalenkoSNF commented on src/django/api/views/v1/moderation_events.py:
```diff
@@ -0,0 +1,60 @@
+class ModerationEvents(ViewSet):
+    swagger_schema = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.opensearch_service = OpenSearchService()
+        self.moderation_events_query_builder = ModerationEventsQueryBuilder()
+        self.opensearch_query_director = OpenSearchQueryDirector(
+                self.moderation_events_query_builder
+            )
```
Whre is `setup()` method here?

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/request_type_validator.py:
```diff
@@ -0,0 +1,22 @@
+        request_type = data.get('request_type')
+
+        if not request_type:
+            return errors
```
request_type is not required.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/date_range_validator.py:
```diff
@@ -0,0 +1,21 @@
+        date_gte = data.get('date_gte')
+        date_lt = data.get('date_lt')
+
+        if date_gte and date_lt and date_gte > date_lt:
+            errors.append({
+                "field": "date_gte",
+                "message": (
+                    "The 'date_gte' must be "
+                    "less than or equal to 'date_lt'."
+                )
+            })
```
We don't need this validation here (and throwing errors on validators level) because we want to collect error messages later by writing them into `errors` list.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/date_range_validator.py:
```diff
@@ -0,0 +1,21 @@
+    def validate_opensearch_params(self, data):
+        errors = []
+
+        date_gte = data.get('date_gte')
+        date_lt = data.get('date_lt')
+
```
We don't have a need to check this here.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -123,26 +99,40 @@ def __build_os_id(self, values):
+                else:
+                    self.query_body['query']['bool']['must'].append({
+                        'range': {field: range_query}
+                    })
```
This is builder pattern, director call API of query builder instance here.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/moderation_events_query_builder.py:
```diff
@@ -0,0 +1,103 @@
+        There should always be sort if there is a search_after field.
+        So if it is empty, sort by name by default
+        '''
+        if not self.query_body['sort']:
```
Fixed comment.

VadimKovalenkoSNF commented on src/django/api/views/v1/moderation_events.py:
```diff
@@ -0,0 +1,60 @@
+                    )
+                },
+                status=status.HTTP_401_UNAUTHORIZED
+            )
```
Fixed.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/request_type_validator.py:
```diff
@@ -0,0 +1,22 @@
+                    Allowed values are 'CREATE', 'UPDATE' or 'CLAIM'."
+            })
+
+        return errors
```
We don't validate type here, we check request type list choices.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/moderation_events_query_builder.py:
```diff
@@ -0,0 +1,103 @@
+                }
+            )
+
+    def _add_search_after(self, search_after):
+        # search_after can't be present as empty by default in query_body
+        if V1_PARAMETERS_LIST.SEARCH_AFTER not in self.query_body:
+            self.query_body[V1_PARAMETERS_LIST.SEARCH_AFTER] = []
+        '''
+        There should always be sort if there is a search_after field.
+        So if it is empty, sort by created_at by default
+        '''
+        if not self.query_body['sort']:
+            sort_criteria = {
+                self.default_sort: {
+                    'order': self.default_sort_order
+                }
+            }
+            self.query_body['sort'].append(sort_criteria)
+
+        self.query_body[V1_PARAMETERS_LIST.SEARCH_AFTER].append(search_after)
```
We don't raise errors on query builder level.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/moderation_events_query_builder.py:
```diff
@@ -0,0 +1,103 @@
+    def _build_country(self, field):
+        return f'cleaned_data.{field}.alpha_2'
+
+    def _build_date_range(self, query_params):
+        date_start = query_params.get('date_gte')
+        date_end = query_params.get('date_lt')
+        range_query = {}
+
+        if date_start is not None:
+            range_query['gte'] = date_start
+        if date_end is not None:
+            range_query['lte'] = date_end
+
+        if range_query:
+            existing_range = any(
+                query.get('range', {}).get('created_at')
+                for query in self.query_body['query']['bool']['must']
+            )
+            if not existing_range:
+                self.query_body['query']['bool']['must'].append({
+                    'range': {'created_at': range_query}
+                })
+
```
We don't validate types here. Do not rename terminology as it is already defined in specs.

VadimKovalenkoSNF commented on src/django/api/views/v1/moderation_events.py:
```diff
@@ -0,0 +1,47 @@
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        query_body = self.opensearch_query_director.build_query(
+            request.GET
+        )
```
We do this in `src/django/api/serializers/v1/moderation_events_serializer.py`.

roman-stolar commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,244 @@
+from django.api.models.moderation_event import ModerationEvent
```
We couldn't used django module in integration tests

VadimKovalenkoSNF replied:
Fixed! It was initially coderabbit suggestion and it was wrong.

VadimKovalenkoSNF commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,246 @@
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        dates = [event['created_at'] for event in result['data']]
+        self.assertEqual(dates, sorted(dates))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_created_at_descending(self):
+        query = "?sort_by=created_at&order_by=desc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        dates = [event['created_at'] for event in result['data']]
+        self.assertEqual(dates, sorted(dates, reverse=True))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_status_ascending(self):
+        query = "?sort_by=status&order_by=asc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        statuses = [event['status'] for event in result['data']]
+        self.assertEqual(statuses, sorted(statuses))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_status_descending(self):
+        query = "?sort_by=status&order_by=desc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        statuses = [event['status'] for event in result['data']]
+        self.assertEqual(statuses, sorted(statuses, reverse=True))
+        self.assertEqual(response.status_code, 200)
+
```
We can't import  ModerationEvent here as this is an integration test that runs in separate from Django containder.

VadimKovalenkoSNF commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,244 @@
+        self.assertEqual(error['field'], 'Country')
+        self.assertEqual(error['message'], "'Usa' Is Not A Valid Alpha-2 Country Code.")
+
+    def test_valid_moderation_id(self):
+        valid_moderation_id = '3b50d60f-85b2-4a17-9f8d-5d3e1fc68198'
+        query = f"?moderation_id={valid_moderation_id}"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+
+        self.assertEqual(response.status_code, 200)
```
We don't have access to ModerationEvents API in the integration tests because it is running in a separate container.

VadimKovalenkoSNF commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,244 @@
+        self.assertEqual(response.status_code, 200)
+
+    def test_filter_by_status(self):
+        status = "APPROVED"
+        query = f"?status={status}"
```
Again, we can't import ModerationEvent here as this is an integration test that runs in separate from Django containder.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/production_locations_query_builder.py:
```diff
@@ -0,0 +1,62 @@
+        There should always be sort if there is a search_after field.
+        So if it is empty, sort by name by default
+        '''
+        if not self.query_body['sort']:
+            sort_criteria = {
+                f'{self.default_sort}.keyword': {
+                    'order': self.default_sort_order
+                }
+            }
+            self.query_body['sort'].append(sort_criteria)
+
```
I don't think it is necessary here, especially when we try to keep API of query builder clear and avoid helper methods there as much as possible.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_director.py:
```diff
@@ -67,16 +74,19 @@ def build_query(self, query_params):
             order_by = query_params.get(V1_PARAMETERS_LIST.ORDER_BY)
-            self.__builder.add_sort(sort_by, order_by)
+            self.__builder._add_sort(sort_by, order_by)
 
         search_after = query_params.get(V1_PARAMETERS_LIST.SEARCH_AFTER)
         if search_after:
-            self.__builder.add_search_after(search_after)
+            self.__builder._add_search_after(search_after)
+
+        paginate_from = query_params.get(V1_PARAMETERS_LIST.FROM)
+        if paginate_from:
+            self.__builder.add_from(paginate_from)
```
We don't do validation here. Instead we have `opensearch_common_validators` folder with necessary validators.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_director.py:
```diff
@@ -22,14 +22,21 @@ def __init__(self, builder):
             V1_PARAMETERS_LIST.AFFILIATIONS: 'terms',
             V1_PARAMETERS_LIST.CERTIFICATIONS_STANDARDS_REGULATIONS: 'terms',
             V1_PARAMETERS_LIST.COORDINATES: 'geo_distance',
+            V1_PARAMETERS_LIST.CONTRIBUTOR_ID: 'terms',
+            V1_PARAMETERS_LIST.REQUEST_TYPE: 'terms',
+            V1_PARAMETERS_LIST.SOURCE: 'terms',
+            V1_PARAMETERS_LIST.STATUS: 'terms',
+            V1_PARAMETERS_LIST.MODERATION_ID: 'terms',
+            V1_PARAMETERS_LIST.DATE_GTE: 'range',
+            V1_PARAMETERS_LIST.DATE_LT: 'range',
```
We do sanitizing in related serializers and validators.

VadimKovalenkoSNF commented on src/django/api/views/v1/moderation_events.py:
```diff
@@ -0,0 +1,47 @@
+            request.GET
+        )
+
+        response = self.opensearch_service.search_index(
+            OpenSearchIndexNames.MODERATION_EVENTS_INDEX,
+            query_body
+        )
+        return Response(response)
```
You suggestnion with `next` and `previous` is good but doens't conform our specs and out of the current scope of work.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/status_validator.py:
```diff
@@ -0,0 +1,31 @@
+                "field": "status",
+                "message": "Status must be a list of values."
+            })
+        elif not all(item in self.VALID_STATUSES for item in status):
+            invalid_statuses = [item for item in status if
+                                item not in self.VALID_STATUSES]
+            errors.append({
```
We do this in Model itself.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/moderation_id_validator.py:
```diff
@@ -0,0 +1,44 @@
+        if moderation_id is None:
+            return errors
+
+        if isinstance(moderation_id, list):
+            invalid_ids = [
+                mid for mid in moderation_id
+                if not re.match(self.UUID_REGEX, mid)
+            ]
+            if invalid_ids:
```
This is out of the scope. Moreover, we have limit of page size that OpenSearch should respond with.

VadimKovalenkoSNF commented on src/django/api/serializers/v1/opensearch_common_validators/moderation_id_validator.py:
```diff
@@ -0,0 +1,44 @@
+import re
+from api.serializers.v1.opensearch_validation_interface \
+    import OpenSearchValidationInterface
+
```
Could you provide a full example of how to apply `uuid` module here?

VadimKovalenkoSNF commented on src/django/api/tests/test_moderation_events_query_builder.py:
```diff
@@ -0,0 +1,110 @@
+            []
+        )
+
+    def test_add_date_range(self):
+        self.builder._build_date_range({
+            'date_gte': '2023-01-01',
+            'date_lt': '2023-12-31'
+        })
+        expected = {
+            'range': {
+                'created_at': {
+                    'gte': '2023-01-01',
+                    'lte': '2023-12-31'
+                }
+            }
+        }
+        self.assertIn(
+            expected,
+            self.builder.query_body['query']['bool']['must']
+        )
```
`lte` is internal OpenSearch property for range queries. And `date_lt` is what we have in specs and prefer to keep it as is.

VadimKovalenkoSNF commented on src/django/api/tests/test_moderation_events_query_builder.py:
```diff
@@ -0,0 +1,110 @@
+            self.builder.query_body['query']['bool']['must']
+        )
+
+    def test_add_sort(self):
+        self.builder._add_sort('created_at', 'asc')
+        expected = {'created_at': {'order': 'asc'}}
+        self.assertIn(expected, self.builder.query_body['sort'])
+
+    def test_add_search_after(self):
+        self.builder._add_search_after('test')
+        self.assertIn('test', self.builder.query_body['search_after'])
+        self.assertIn(
+            {'created_at': {'order': 'desc'}},
+            self.builder.query_body['sort']
+        )
```
We don't raise `ValueError`. Instead we throw `ValidationError` like:
```
if errors:
  # [OSDEV-1441] Pass error msg to the Rollbar here
  raise ValidationError({
      "message": COMMON_ERROR_MESSAGE,
      "errors": errors
  })
```

As a result we will have an error-like response with status code 400.

VadimKovalenkoSNF replied:
These assertion won't work neither, because you create test queries directly in OpenSearch builder but our software assumed that these params were validated on model level before passing here.

VadimKovalenkoSNF commented on src/django/api/tests/test_moderation_events_query_builder.py:
```diff
@@ -0,0 +1,110 @@
+            []
+        )
+
+    def test_add_date_range(self):
+        self.builder._build_date_range({
+            'date_gte': '2023-01-01',
+            'date_lt': '2023-12-31'
+        })
+        expected = {
+            'range': {
+                'created_at': {
+                    'gte': '2023-01-01',
+                    'lte': '2023-12-31'
+                }
+            }
+        }
+        self.assertIn(
+            expected,
+            self.builder.query_body['query']['bool']['must']
+        )
```
We don't raise ValueErrros for these parameters.

roman-stolar reviewed (dismissed):
In general LGTM

VadimKovalenkoSNF commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,243 @@
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        dates = [event['created_at'] for event in result['data']]
+        self.assertEqual(dates, sorted(dates))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_created_at_descending(self):
+        query = "?sort_by=created_at&order_by=desc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        dates = [event['created_at'] for event in result['data']]
+        self.assertEqual(dates, sorted(dates, reverse=True))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_status_ascending(self):
+        query = "?sort_by=status&order_by=asc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        statuses = [event['status'] for event in result['data']]
+        self.assertEqual(statuses, sorted(statuses))
+        self.assertEqual(response.status_code, 200)
+
+    def test_sort_by_status_descending(self):
+        query = "?sort_by=status&order_by=desc"
+        response = requests.get(
+            f"{self.root_url}/api/v1/moderation-events/{query}",
+            headers=self.basic_headers,
+        )
+        result = response.json()
+        statuses = [event['status'] for event in result['data']]
+        self.assertEqual(statuses, sorted(statuses, reverse=True))
+        self.assertEqual(response.status_code, 200)
+
```
I'm fine with current implementation.

VadimKovalenkoSNF commented on src/tests/v1/test_moderation_events.py:
```diff
@@ -0,0 +1,243 @@
+            headers=self.basic_headers,
+        )
+        self.assertEqual(response.status_code, 200)
+
```
Done.

roman-stolar reviewed (dismissed):
Re-Approved.

mazursasha1990 commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -123,26 +99,41 @@ def __build_os_id(self, values):
         )
 
+    @abstractmethod
+    def _build_date_range(self, query_params):
```
Why this method should have the @abstractmethod decorator, as it is only used in the ModerationEventsQueryBuilder?
In my opinion, the definition of the _build_date_range method should be moved into the OpenSearchQueryBuilder and it should be marked as private.

VadimKovalenkoSNF replied:
> In my opinion, the definition of the _build_date_range method should be moved into the OpenSearchQueryBuilder and it should be marked as private.

I agree with you. Fixed. Thank you for pointing out this.

mazursasha1990 reviewed (changes requested):
Please pay attention to the opensearch query builder classes. Fix abstract methods and access modifiers for different methods.

Here is an article about [Access Modifiers in Python : Public, Private and Protected](https://www.geeksforgeeks.org/access-modifiers-in-python-public-private-and-protected/)

VadimKovalenkoSNF commented on doc/release/RELEASE-NOTES.md:
```diff
@@ -4,6 +4,36 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html). The format is based on the `RELEASE-NOTES-TEMPLATE.md` file.
 
 
+## Release 1.24.0
```
Fixed.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -66,9 +51,12 @@ def __build_number_of_workers(self, field, range_query):
         })
 
-    def __build_country(self, field):
+    def _build_country(self, field):
```
> Moreover, it should be marked as private.
Agreed, I made it a private for `ModerationEventsQueryBuilder` and `ProductionLocationsQueryBuilder` separately.


> It should be an abstract method and defined in the OpenSearchQueryBuilderInterface, because the implementation of the _build_country is different in ModerationEventsQueryBuilder and ProductionLocationsQueryBuilder.

In Python, abstract base classes (ABCs) can contain private methods, but generally, private methods aren't declared abstract because the purpose of an abstract method is to enforce an interface on subclasses.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -1,30 +1,15 @@
         self.query_body = copy.deepcopy(self.default_query_body)
 
-    def __build_number_of_workers(self, field, range_query):
+    def _build_number_of_workers(self, field, range_query):
```
Good point. Done.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder_interface.py:
Done, pls check.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -94,23 +82,11 @@ def add_multi_match(self, query):
-                field, lambda x: f'{x}.keyword'
-            )(field)
+    @abstractmethod
+    def _add_terms(self, field, values):
```
It should be here to follow inheritance principles, but I refactored it.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_director.py:
```diff
@@ -67,16 +74,19 @@ def build_query(self, query_params):
         if sort_by:
             order_by = query_params.get(V1_PARAMETERS_LIST.ORDER_BY)
-            self.__builder.add_sort(sort_by, order_by)
+            self.__builder._add_sort(sort_by, order_by)
```
Yes! I re-worked methods that started from `add_...` so they all public in the client code.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_director.py:
```diff
@@ -67,16 +74,19 @@ def build_query(self, query_params):
         search_after = query_params.get(V1_PARAMETERS_LIST.SEARCH_AFTER)
         if search_after:
-            self.__builder.add_search_after(search_after)
+            self.__builder._add_search_after(search_after)
```
Agreed, fixed.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -153,30 +144,13 @@ def add_geo_distance(self, field, lat, lng, distance):
-            }
-            self.query_body['sort'].append(sort_criteria)
+    @abstractmethod
+    def _add_sort(self, field, order_by=None):
```
I have to keep it here, it will be called from subclasses (`ModerationEventsQueryBuilder` and `ProductionLocationsQueryBuilder`).

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -153,30 +144,13 @@ def add_geo_distance(self, field, lat, lng, distance):
 
-        self.query_body[V1_PARAMETERS_LIST.SEARCH_AFTER].append(search_after)
+    @abstractmethod
+    def _add_search_after(self, search_after):
```
I have to keep it here, it will be called from subclasses (ModerationEventsQueryBuilder and ProductionLocationsQueryBuilder).

VadimKovalenkoSNF commented:
@mazursasha1990 Thanks for the comprehensive review. It encouraged me to look at the inheritance problem here from another angle. There current query builder inheritance will follow this pattern after my recent changes:

```
from abc import ABC, abstractmethod

class OpenSearchQueryBuilderInterface(ABC):
    @abstractmethod
    def add_sort(self, field, order_by=None):
        """Public method for adding sort functionality."""
        pass

class OpenSearchQueryBuilder(OpenSearchQueryBuilderInterface):
    """Base class that provides common query-building utilities."""

    def add_sort(self, field, order_by=None):
        """Public interface that calls the protected `_add_sort` method."""
        self._add_sort(field, order_by)

    @abstractmethod
    def _add_sort(self, field, order_by=None):
        """Protected method that subclasses can customize."""
        pass

class ModerationEventsQueryBuilder(OpenSearchQueryBuilder):
    """Concrete implementation of the OpenSearchQueryBuilder for moderation events."""

    def _add_sort(self, field, order_by=None):
        # Concrete implementation specific to moderation events
        print(f"Sorting {field} in {order_by} order for moderation events.")

```

In a nutshell, this allows to call `add_...` methods as public without breaking abstraction in  `OpenSearchQueryBuilderInterface` -> `OpenSearchQueryBuilder` ->  `ModerationEventsQueryBuilder` | `ProductionLocationsQueryBuilder`

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -123,27 +58,58 @@ def __build_os_id(self, values):
+                    'range': {'created_at': range_query}
                 })
 
+    def add_range(self, field, query_params):
+        if field in {
+            V1_PARAMETERS_LIST.NUMBER_OF_WORKERS,
+            V1_PARAMETERS_LIST.PERCENT_FEMALE_WORKERS
+        }:
+            min_value = query_params.get(f'{field}[min]')
+            max_value = query_params.get(f'{field}[max]')
+
+            min_value = int(min_value) if min_value else None
+            max_value = int(max_value) if max_value else None
+
+            range_query = {}
+            if min_value is not None:
+                range_query['gte'] = min_value
+            if max_value is not None:
+                range_query['lte'] = max_value
+            if range_query:
+                build_action = self.build_options.get(field)
+
+                if build_action:
+                    build_action(field, range_query)
+                else:
+                    self.query_body['query']['bool']['must'].append({
+                        'range': {field: range_query}
+                    })
+
+        elif field in {
+            V1_PARAMETERS_LIST.DATE_GTE,
+            V1_PARAMETERS_LIST.DATE_LT
+        }:
+            self.__build_date_range(query_params)
```
Data already validated on different (upper) level.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -123,27 +58,58 @@ def __build_os_id(self, values):
-        if max_value is not None:
-            range_query['lte'] = max_value
+
+        if date_start is not None:
+            range_query['gte'] = date_start
+        if date_end is not None:
+            range_query['lte'] = date_end
 
         if range_query:
-            build_action = self.build_options.get(field)
-            if build_action:
-                build_action(field, range_query)
-            else:
+            existing_range = any(
+                query.get('range', {}).get('created_at')
+                for query in self.query_body['query']['bool']['must']
+            )
+            if not existing_range:
                 self.query_body['query']['bool']['must'].append({
-                    'range': {field: range_query}
+                    'range': {'created_at': range_query}
                 })
```
Data already validated on different (upper) level.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -1,73 +1,16 @@
-
-    def __build_country(self, field):
-        return f'{field}.alpha_2'
+    def add_from(self, paginate_from):
+        self.query_body['from'] = paginate_from
```
We dont' check this here. Data already validated on different (upper) level.

VadimKovalenkoSNF commented:
UPD: OpenSearch query builder inheritance has been refactored again. I removed `OpensearchQueryBuilderInterface`. Instead use `ABC` for `class OpenSearchQueryBuilder(ABC)` from `abc` module. Also, for sake of readability, I removed protected methods signatures at all. Current logic for query builders is:

`OpenSearchQueryBuilder(ABC)` -> `ModerationEventsQueryBuilder` | `ProductionLocationsQueryBuilder`.
Also, restructured methods order in `OpenSearchQueryBuilder(ABC)`.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_builder.py:
```diff
@@ -153,30 +96,31 @@ def add_geo_distance(self, field, lat, lng, distance):
     def get_final_query_body(self):
         return self.query_body
+
+    # Call this method in child classes
+    def __build_os_id(self, values):
+        # Build a query to search in both os_id and historical_os_id.keyword
+        self.query_body['query']['bool']['must'].append(
+            {
+                'bool': {
+                    'should': [
+                        {'terms': {'os_id': values}},
+                        {'terms': {'historical_os_id.keyword': values}},
+                    ]
+                }
+            }
+        )
```
This should be private to follow consideration that all `__build_...` should be private. Child classes invokes this method by getting access to parent class first.

roman-stolar reviewed (approved):
LGTM

mazursasha1990 reviewed (changes requested):
In general, looks good to me.
But I strongly recommend to escape from name-mangled method from a subclass and use protected (single underscore) method.
My thoughts I explained in the comment.

mazursasha1990 reviewed (approved):
LGTM

VadimKovalenkoSNF commented:
@mazursasha1990 I made `_build_os_id` as protected, pls re-check.
